README.md
# whisper-asr

Test whister as ASR.

## Prerequisites

Python >= 3.8

### on MacOS using Homebrew (<https://brew.sh/>)

``` shell
brew install ffmpeg
brew install portaudio
```

### Create Virtual Environment

``` shell
python -m venv venv
echo 'venv' > .gitignore
```

### Activate and Deactivate Virtual Environment

``` shell
source venv/bin/activate
```

``` shell
source deactivate
```

## Installation

``` shell
pip install -r requirements.txt
```

Backlog:

``` shell
pip install -U openai-whisper
pip install setuptools-rust
pip install pyaudio
```

### Freeze the requirements

``` shell
pip freeze > requirements.txt
```

## Example

### Run

``` shell
python example.py --audio_file_name=german           
Detected language: de
Detected text: Wie ist das weder ein Passwort?
Translated transcribt: How is the WLAN Password?
```

## Whisper API

### Enviroment Variables

[Whisper API](https://platform.openai.com/docs/api-reference/audio) needs API key from OpenAI as enviroment variables in `.env`:

``` shell
export OPENAI_API_KEY =
```

### Call API

``` shell
source .env
python whisper-api.py
python whisper-api.py --audio_file_name english --methode transcript  
python whisper-api.py --methode translate
```

## Server-Client

### run Server

``` shell
flask --app server run --port 5001
```

### run Client

``` shell
python client.py
python client.py --audio_file_name english
```

``` shell
curl -F "file=@audio/deutsch.mp3" http://localhost:5001/whisper
```

## Dockerize

``` shell
docker build . -t leonardmf/whisper-asr:0.2.1 -t leonardmf/whisper-asr:latest
docker run -p 5001:5000 leonardmf/whisper-asr:latest
docker run -p 8765:8765 leonardmf/whisper-asr:0.2.1

# with GPU
docker run --gpus=all -p 8765:8765 leonardmf/whisper-asr:0.2.1
docker run --gpus=all -p 5001:5000 leonardmf/whisper-asr:latest
```

``` shell
docker push leonardmf/whisper-asr:0.2.1
docker push leonardmf/whisper-asr:latest
```

``` shell
curl -F "file=@audio/deutsch.mp3" http://localhost:5001/whisper
```

## Audio

``` shell
ffmpeg -i input.m4a -vn -ar 44100 -ac 2 -b:a 192k output.mp3
```

## Streaming Server-Client

ASR with streaming and end-of-speech detection.

In terminal 1:

``` shell
python streaming_server.py
WebSocket server started. Listening on port 8765.
# Or docker
docker run -p 8765:8765 leonardmf/whisper-asr:0.2.1
# with GPU
docker run --gpus=all -p 8765:8765 leonardmf/whisper-asr:0.2.1
```

In terminal 2:

``` shell
python streaming_client.py
python streaming_client.py --task translate
WebSocket connection established.
Start streaming audio ...
```

Start to speak ...


CHANGELOG.md
# Whisper ASR Server-Prototyp Release Notizen


## 0.4.0.0006 Alpha (05.07.2023)

Features:

* Lokales PyTorch Whisper-Modell erstellt
* Ausbau auf 54 Audio-Dateien zum Training
* Aufbau des Trainings von "Ahoi Stella" im Multilanguage Tiny-Modell von Whisper
* Erstellen der Whisper-KWS Version zur Püfung von "Ahoi Stella" als Wakeword


## 0.3.1.0005 Alpha (10.06.2023)

Features: 

* Versionsanzeige eingebaut
* Logging eingebaut
* Aufbau des Whisper-Python Servers für die Prüfung eines Wakewords
* Aufteilung der Docker Container in Whisper-Base und Whisper-Asr, um unnötiges Laden zu vermeiden


## 0.3.0.0004 Alpha (01.06.2023)

Features:

* whisper_cpp_server.py eingefügt, um pywhisptercpp und whisper.cpp zu verwenden



streaming_client.py
import asyncio
import websockets
import time
import sounddevice as sd
import numpy as np
import asyncclick as click

LOCAL_WEB_SOCKET_URL = "ws://127.0.0.1:8765"
# LOCAL_WEB_SOCKET_URL = "ws://192.168.178.123:8765"
# DEFAULT_SAMPLE_RATE = "44100"
DEFAULT_SAMPLE_RATE = "16000"
DEFAULT_CHANNELS = "1"
DEFAULT_DTYPE = "int16"
# DEFAULT_TASK = "translate"
DEFAULT_TASK = "transcribe"


async def audio_stream(
    samplerate = DEFAULT_SAMPLE_RATE,
    channels   = DEFAULT_CHANNELS,
    dtype      = DEFAULT_DTYPE,
    task       = DEFAULT_TASK,
    url        = LOCAL_WEB_SOCKET_URL,
):
    samplerate = int(samplerate)
    channels   = int(channels)
    
    # Set the WebSocket request headers
    headers = {
        "samplerate": samplerate,
        "channels": channels,
        "dtype": dtype,
        "task": task
    }
    
    # Open the WebSocket connection
    async with websockets.connect(url,extra_headers=headers) as websocket:
        print("WebSocket connection established.")
    
        # Configure sounddevice input stream
        stream = sd.InputStream(
            channels=channels,
            samplerate=samplerate,
            dtype=dtype,
        )
        
        # Start recording audio
        stream.start()
        print("Start streaming audio ...")
        
        start_of_speech_flag = False
        start_time = time.time()
        
        while True:
            # Read audio data from input stream
            audio_data, _ = stream.read(1024)
            
            # Detect start of speech
            if np.mean(audio_data) > 10:
                start_of_speech_flag = True 
                start_time = time.time()
    
            # Detect end of speech
            if start_of_speech_flag and time.time() - start_time >= 0.55:
                stream.stop()
                print("Stop streaming audio ...")
                # send end of speech
                eos_time = time.time()
                await websocket.send("end of speech")
                break
            else: 
                # Send audio data to the WebSocket server
                await websocket.send(audio_data.tobytes())
         
        # Receive any response from the WebSocket server (if needed)
        response = await websocket.recv()
        response_time = time.time()
        print(response)
        print("Response time: " + str(response_time - eos_time))
        await websocket.close()
        
@click.command()
@click.option(
    "--task",
    default="transcribe",
    type=str,
    help="Task to perform: translate or transcribe",
)
@click.option(
    "--url",
    default=LOCAL_WEB_SOCKET_URL,
    type=str,
    help="URL of the WebSocket server",
) 
@click.option(
    "--samplerate",
    default="44100",
    type=str,
    help="Sample rate of the audio stream",
)
@click.option(
    "--channels",
    default="1",
    type=str,
    help="Channels of the audio stream",
)       
async def main(task: str, url: str, samplerate: str, channels: str):
    # Run the audio_stream function in an event loop
    await audio_stream(task=task, url=url, samplerate=samplerate, channels=channels)

if __name__ == "__main__":
    asyncio.run(main())
